### PR TITLE
Rsgraber/20230508 adr

### PR DESCRIPTION
**Pull request by rgraber**
Add an ADR about how and why we will be allowing multiple event types per topic with the event bus.

Note:
While the ADR obviously posits a decision as made, I am somewhat ambivalent on the proposed solution given the consequences. I would definitely appreciate other opinions, especially as regards either using a SettingsToggle or using Avro unions.

**Merge checklist:**
Check off if complete *or* not applicable:
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Commits are squashed
- [ ] Noted any: Concerns, dependencies, deadlines, tickets, testing instructions

### CODE REVIEW

File: docs/decisions/0011_multiple_event_types_per_topic.rst [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+    {
+        "doc": "Avro Event Format for CloudEvents created with openedx_events/schema",
+        "fields": [{"name": "str_field", "type": "string"}]
+        "name": "CloudEvent",
+        "namespace": "my.event.type",
+        "type": "record"
+    }
+
+and the topic ``my-topic``, the serializer will look for a schema registered as ``my-topic-my.event.type.CloudEvent`` and use that to determine if the schema being passed is valid. This means that as long as the namespace is different for the schemas for different event types, the serializer will not error when adding new event types. It will still maintain evolution rules within an event type, however, which is desirable behavior.
```
Comment (robrap) [inline docs/decisions/0011_multiple_event_types_per_topic.rst:39]: It's unclear how exactly `record-name` relates to the `name` field and `namespace` field. Are these both standard fields looked at by something (what?), and is `namespace` used if provided, otherwise `name` is used?

File: docs/decisions/0011_multiple_event_types_per_topic.rst [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+Consequences
+************
+* The record names for each event type will need to be distinct from each other. Originally, they were all just "CloudEvent." We needed to add namespaces to ensure that they were unique. Note that this will be necessary regardless of which solution we choose.
+* Any event type will be able to be written to any topic
```
Comment (robrap) [inline docs/decisions/0011_multiple_event_types_per_topic.rst:44]: Maybe something like following, to make it clear that this is an unintended consequence?
```suggestion
* Any event type will be able to be written to any topic, even for topics where we only want a single event type.
```

File: docs/decisions/0011_multiple_event_types_per_topic.rst [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+However, other types of events are not so easily grouped into a single signal. For example, there are different signals for ``XBLOCK_PUBLISHED``, ``XBLOCK_UPDATED``, and ``XBLOCK_DUPLICATED``, and indeed different schemas across the three signals. Routing these signals through different topics could lead to events being processed in a nonsensical order, for example an xblock being duplicated before it's been published. See `Should You Put Several Event Types in the Same Kafka Topic?`_ and `Putting Several Event Types in the Same Topic – Revisited`_ for more information on why it can be useful to group different event types on the same topic.
+
+.. _Should You Put Several Event Types in the Same Kafka Topic?: https://www.confluent.io/blog/put-several-event-types-kafka-topic/
+.. _Putting Several Event Types in the Same Topic – Revisited: https://www.confluent.io/blog/multiple-event-types-in-the-same-kafka-topic/
+
+Decision
+********
+We will update the code for event serializers, which are responsible for registering and validating schemas against the schema registry, to use the TopicRecordName strategy.
```
Comment (robrap) [inline docs/decisions/0011_multiple_event_types_per_topic.rst:20]: Maybe start with something like:

> We will allow multiple related event types to be produced to a single topic so that they can be ordered appropriately. We will accomplish this by...

File: docs/decisions/0011_multiple_event_types_per_topic.rst [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        "type": "record"
+    }
+
+and the topic ``my-topic``, the serializer will look for a schema registered as ``my-topic-my.event.type.CloudEvent`` and use that to determine if the schema being passed is valid. This means that as long as the namespace is different for the schemas for different event types, the serializer will not error when adding new event types. It will still maintain evolution rules within an event type, however, which is desirable behavior.
+
+Consequences
+************
+* The record names for each event type will need to be distinct from each other. Originally, they were all just "CloudEvent." We needed to add namespaces to ensure that they were unique. Note that this will be necessary regardless of which solution we choose.
+* Any event type will be able to be written to any topic
+* Schema evolution rules will be enforced for event types but not for topics.
```
Comment (robrap) [inline docs/decisions/0011_multiple_event_types_per_topic.rst:45]: I wonder if "but not for topics" would be a little confusing for non-experts? What about something like:
```suggestion
* Schema evolution rules will be enforced separately for each event type on a topic.
```

File: docs/decisions/0011_multiple_event_types_per_topic.rst [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+Using different serializers based on the topic
+==============================================
+We could refactor the code to configure the AvroSerializer differently based on the topic, only changing the subjec name strategy when the topic should allow multiple event types. This would be a significant amount of work and would require some list of which topics accept multiple event types, which would have to be maintained and would reduce the flexibility of the event bus.
+
+Require related event types to have the same schema
+===================================================
+In theory we could avoid the issue of event types with different schemas on the same topic by simply not allowing them, requiring all event types that are intended to go on the same topic to have the same schema. This would require knowing in advance which event types will go to which topics and likely result in lots of extraneous fields which are only necessary for some of the event types and not others.
```
Comment (robrap) [inline docs/decisions/0011_multiple_event_types_per_topic.rst:66]: Also, this is considered an anti-pattern, so we should definitely avoid this.

File: docs/decisions/0011_multiple_event_types_per_topic.rst [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+and the topic ``my-topic``, the serializer will look for a schema registered as ``my-topic-my.event.type.CloudEvent`` and use that to determine if the schema being passed is valid. This means that as long as the namespace is different for the schemas for different event types, the serializer will not error when adding new event types. It will still maintain evolution rules within an event type, however, which is desirable behavior.
+
+Consequences
+************
+* The record names for each event type will need to be distinct from each other. Originally, they were all just "CloudEvent." We needed to add namespaces to ensure that they were unique. Note that this will be necessary regardless of which solution we choose.
+* Any event type will be able to be written to any topic
+* Schema evolution rules will be enforced for event types but not for topics.
+* The event-bus-kafka consumer will need to be updated to determine the signal from the message headers rather than taking a signal as a passed argument in the management command. This will also be necessary regardless of which solution we choose.
+
```
Comment (robrap) [inline docs/decisions/0011_multiple_event_types_per_topic.rst:47]: Somewhere could we note that in the future we might be able to detect if a service is only sending one event type for topic for all topics, and to not configure in the same way in this case?

**Review by robrap — COMMENTED** [review]
I wonder if we need an ADR in openedx-events simply stating that all implementations (including Redis as an example) will need to be able to support sending multiple event types to a single topic to allow for better support of ordering these events. This ADR would then just be about how we implement this other decision in Kafka. Thoughts?

**Comment by rgraber** [comment]
> I wonder if we need an ADR in openedx-events simply stating that all implementations (including Redis as an example) will need to be able to support sending multiple event types to a single topic to allow for better support of ordering these events. This ADR would then just be about how we implement this other decision in Kafka. Thoughts?

I think that's a good idea. I should probably hold off on this ADR so it can point to that one. I'll move some of the text in the context section to the openedx-events ADR.

File: docs/decisions/0011_multiple_event_types_per_topic.rst (hunk 3 above)
Comment (rgraber) [reply docs/decisions/0011_multiple_event_types_per_topic.rst]: [resolved]

File: docs/decisions/0011_multiple_event_types_per_topic.rst (hunk 5 above)
Comment (rgraber) [reply docs/decisions/0011_multiple_event_types_per_topic.rst:65]: [resolved]

File: docs/decisions/0011_multiple_event_types_per_topic.rst (hunk 6 above)
Comment (rgraber) [reply docs/decisions/0011_multiple_event_types_per_topic.rst:46]: I'm not sure that's true. See the consequences of using a rollout toggle. We pretty much need this to be all-or-nothing across services.

File: docs/decisions/0011_multiple_event_types_per_topic.rst [hunk 7]
```diff
@@ -0,0 +1,73 @@
+10. Multiple event types per topic
+##################################
+
+Status
+******
+
+- **Accepted** DATE TO BE ADDED BEFORE MERGING
```
Comment (robrap) [inline docs/decisions/0011_multiple_event_types_per_topic.rst:7]: Nit: You can drop the `-`.

File: docs/decisions/0011_multiple_event_types_per_topic.rst [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+Context
+*******
+
+Following `this ADR`_ in openedx-events, all implementations of event bus must support multiple event types on the same topic. This ADR explains how we will implement this using Kafka.
+**UPDATE LINK AFTER THAT ADR MERGES**
+
```
Comment (robrap) [inline docs/decisions/0011_multiple_event_types_per_topic.rst:14]: Can you add a small section providing a brief list of possible solutions to this problem in Kafka? Could just be bullets, and you can note that they'll be covered in more detail later. But basically, the context is that there are 2 (or 3, etc. depending on how you look at it) ways of accomplishing this, and this ADR is just about the one we selected.

File: docs/decisions/0011_multiple_event_types_per_topic.rst [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+.. _this ADR: https://github.com/openedx/openedx-events/pull/217
+
+Decision
+********
+To comply with the event bus requirements, we will use the TopicRecordName strategy in the event serializers, which are responsible for registering and validating schemas against the schema registry.
```
Comment (robrap) [inline docs/decisions/0011_multiple_event_types_per_topic.rst:19]: Can you summarize why? For example: This method seems to be the simplest solution that solves the problem. (Or whatever you think the summary would be.) You don't really call out the pros/cons of the selected or rejected alternatives, so the why doesn't really jump out at you. Consider adding pros/cons, or reorganizing to make this more clear.

File: docs/decisions/0011_multiple_event_types_per_topic.rst (hunk 6 above)
Comment (robrap) [reply docs/decisions/0011_multiple_event_types_per_topic.rst:46]: Maybe we can have a chat, simply because I don't follow it. Only a single service should ever produce to a topic. I don't think we have an ADR around that, but that is a pattern that seems to be generally accepted. There might be several instances (e.g. EC2 instances, k8s containers) of the same service, but I don't think that is what you are referring to...

As I was writing this, I had a thought that _may_ be what you were trying to say. The above constraint is possibly about an event type, and not a topic. So, one question is whether this ADR should declare that all event types being produced to a shared topic must be sourced from the same service. I say yes.

Does this have anything to do with the point you were trying to make, or do we also need an entirely separate discussion as well?

File: docs/decisions/0011_multiple_event_types_per_topic.rst [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+Rejected Alternatives
+*********************
+
+Setting the subject name strategy on the topic rather than the client
+=====================================================================
+This option is only available for locally-hosted or Dedicated Confluent Cloud clusters, and thus not a feasible solution for the whole community.
+
+Using a SettingsToggle
+======================
+Usually we would put this sort of feature behind a SettingsToggle so deployers could enable it and disable it as they chose. However, because the subject name strategy is set at the client rather than topic or cluster level, all clients who produce to the same topic(s) must have the same subject name strategy set. While the consequences of having multiple clients with different subject name strategies writing to the same topic are not 100% clear, at the very least it would result in a proliferation of confusing schemas in the registry.
```
Comment (robrap) [inline docs/decisions/0011_multiple_event_types_per_topic.rst:57]: I'm not clear yet if this section will stay or not until we discuss on the other comment thread.

**IF** it stays, I wonder if we should:
- Add to the decision that we will **not** provide.a toggle to disable. (may not be necessary)
- Clarify that this is not an alternative to the TopicRecordName Strategy decision, but an implementation detail of that alternative. Maybe this would just move to the decision section altogether, and drop this from here?

File: docs/decisions/0011_multiple_event_types_per_topic.rst [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+Using different serializers based on the topic
+==============================================
+We could refactor the code to configure the AvroSerializer differently based on the topic, only changing the subjec name strategy when the topic should allow multiple event types. This would be a significant amount of work and would require some list of which topics accept multiple event types, which would have to be maintained and would reduce the flexibility of the event bus.
+
+Require related event types to have the same schema
+===================================================
+In theory we could avoid the issue of event types with different schemas on the same topic by simply not allowing them, requiring all event types that are intended to go on the same topic to have the same schema. This would require knowing in advance which event types will go to which topics and likely result in lots of extraneous fields which are only necessary for some of the event types and not others.
```
Comment (robrap) [inline docs/decisions/0011_multiple_event_types_per_topic.rst:65]: Can you either drop this section, or just point to the other ADR that has the updated language?

**Review by robrap — COMMENTED** [review]
Thank you. The idea around pros/cons is non-blocking.

File: docs/decisions/0011_multiple_event_types_per_topic.rst (hunk 6 above)
Comment (rgraber) [reply docs/decisions/0011_multiple_event_types_per_topic.rst:45]: I was thinking about multiple services producing to the same topic. If we are comfortable saying that is not allowed and if you do that you could end up with weird schema problems, that's fine. I would definitely call that out in the ADR though.

File: docs/decisions/0011_multiple_event_types_per_topic.rst (hunk 8 above)
Comment (rgraber) [reply docs/decisions/0011_multiple_event_types_per_topic.rst:13]: [fixed]

File: docs/decisions/0011_multiple_event_types_per_topic.rst (hunk 6 above)
Comment (rgraber) [reply docs/decisions/0011_multiple_event_types_per_topic.rst:45]: [fixed] Added back the SettingsToggle, and noted that this requires one producer per topic

File: docs/decisions/0011_multiple_event_types_per_topic.rst (hunk 10 above)
Comment (rgraber) [reply docs/decisions/0011_multiple_event_types_per_topic.rst:56]: [fixed] removed

File: docs/decisions/0011_multiple_event_types_per_topic.rst (hunk 11 above)
Comment (rgraber) [reply docs/decisions/0011_multiple_event_types_per_topic.rst]: [fixed]

File: docs/decisions/0011_multiple_event_types_per_topic.rst (hunk 7 above)
Comment (rgraber) [reply docs/decisions/0011_multiple_event_types_per_topic.rst]: [fixed]

File: docs/decisions/0011_multiple_event_types_per_topic.rst [hunk 12]
```diff
@@ -0,0 +1,90 @@
+10. Multiple event types per topic
+##################################
+
+Status
+******
+
+**Accepted** DATE TO BE ADDED BEFORE MERGING
```
Comment (robrap) [inline docs/decisions/0011_multiple_event_types_per_topic.rst:7]: Reminder comment so this isn't forgotten.

**Review by robrap — APPROVED** [review]
Thank you!
